tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -106,20 +120,27 @@ func (wc *weakCoin) Get(epoch types.EpochID, round types.RoundID) bool {
 }
 
 func (wc *weakCoin) PublishProposal(ctx context.Context, epoch types.EpochID, round types.RoundID) error {
-	p, err := wc.generateProposal(epoch, round)
+	wc.Log.With().Info("Publishing proposal",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round_id", uint64(round)))
+
+	proposal, err := wc.generateProposal(epoch, round)
 	if err != nil {
 		return err
 	}
 
-	if wc.proposalExceedsThreshold(p) {
+	signedProposal := wc.signProposal(proposal)
+
+	if wc.proposalExceedsThreshold(proposal) {
```

Comment:
@nkryuchkov this should be proposalExceedsThreshold(signature) , isn't it? otherwise we can completely skip generating proposals for the a particular epoch/round combination